### ISSUE TITLE
Segfault when running tests

### ISSUE THREAD

**Issue by EmRowlands** [whoshuu/cpr#495]
I am packaging cpr for Gentoo, but I have run into an issue with the tests: most of them segfault in my environment. I'm not particularly familiar with the source (I'm packaging as a dependency) so I ask here for help.

I don't think it's a missing dependency (everything builds correctly and `ldd` shows nothing obviously wrong), and I don't think it's my CFLAGS.

Log files are below:
Test log: [LastTest.log](https://github.com/whoshuu/cpr/files/5775625/LastTest.log)
Build log/system configuration: [dev-cpp:cpr-1.5.2:20210106-120325.log](https://github.com/whoshuu/cpr/files/5775656/dev-cpp.cpr-1.5.2.20210106-120325.log)

Basic system information:
```
sh dash 0.5.11.2
ld GNU ld (Gentoo 2.34 p6) 2.34.0
distcc 3.3.3 x86_64-pc-linux-gnu [disabled]
app-shells/bash:          5.0_p18::gentoo
dev-util/cmake:           3.17.4-r1::gentoo
sys-devel/binutils:       2.34-r2::gentoo
sys-devel/gcc:            10.2.0-r5::gentoo
sys-devel/gcc-config:     2.3.2-r1::gentoo
sys-devel/make:           4.3::gentoo
sys-kernel/linux-headers: 5.9::gentoo (virtual/os-headers)
sys-libs/glibc:           2.32-r3::gentoo
```

**Comment by COM8** [comment]
Interesting one. Thanks for reporting.
To me it looks like you are using `cpr-1.5.2`.
Could you please try building the master branche of cpr and running all tests on your system?

**Comment by EmRowlands** [comment]
Test log: [LastTest.log](https://github.com/whoshuu/cpr/files/5781698/LastTest.log)
Build log/system configuration: 
[dev-cpp:cpr-9999:20210107-125556.log](https://github.com/whoshuu/cpr/files/5781702/dev-cpp.cpr-9999.20210107-125556.log)

Seems to be the same result.

**Comment by COM8** [comment]
Do you know if there exists any up to date Gentoo live image?

**Comment by EmRowlands** [comment]
Gentoo doesn't really have a live image, but you can find "stage 4" tarballs which can be unzipped and used as a chroot.
See here for more info: https://wiki.gentoo.org/wiki/Stage_tarball#Stage_4

I also did some investigation myself. I'm not really familiar with mongoose, but it appears `mg_bind` is returning a null pointer.

    $ git clone https://github.com/whoshuu/cpr.git
    [...]
    $ cmake -S cpr -B build -DUSE_SYSTEM_CURL=yes -DCMAKE_BUILD_TYPE=Debug -DUSE_SYSTEM_GTEST=yes -G Ninja
    [...]
    $ cd build
    $ ninja -j4 -l4 
    $ gdb bin/get_tests
    (gdb) run
    Starting program: /tmp/tmp.9PHOp0W6Cq/build/bin/get_tests
    [Thread debugging using libthread_db enabled]
    Using host libthread_db library "/lib64/libthread_db.so.1".
    [==========] Running 96 tests from 11 test suites.
    [----------] Global test environment set-up.
    [New Thread 0x7ffff74a1640 (LWP 17711)]
    
    Thread 2 "get_tests" received signal SIGSEGV, Segmentation fault.
    [Switching to Thread 0x7ffff74a1640 (LWP 17711)]
    0x00005555555f3175 in mg_set_protocol_http_websocket (nc=0x0) at _deps/mongoose-src/mongoose.c:6897
    6897      nc->proto_handler = mg_http_handler;
    (gdb) bt
    #0  0x00005555555f3175 in mg_set_protocol_http_websocket (nc=0x0) at _deps/mongoose-src/mongoose.c:6897
    #1  0x00005555555ccbfd in cpr::HttpServer::initServer (this=0x5555556494c0, mgr=0x555555649568, event_handler=0x5555555ca35b <cpr::EventHandler(mg_connection*, int, void*)>)
        at /tmp/tmp.9PHOp0W6Cq/cpr/test/httpServer.cpp:19
    #2  0x00005555555ca447 in cpr::AbstractServer::Run (this=0x5555556494c0) at /tmp/tmp.9PHOp0W6Cq/cpr/test/abstractServer.cpp:65
    #3  0x00005555555cca28 in std::__invoke_impl<void, void (cpr::AbstractServer::*)(), cpr::AbstractServer*> (
        __f=@0x5555556553a0: (void (cpr::AbstractServer::*)(cpr::AbstractServer * const)) 0x5555555ca3f6 <cpr::AbstractServer::Run()>, __t=@0x555555655398: 0x5555556494c0)
        at /usr/lib/gcc/x86_64-pc-linux-gnu/10.2.0/include/g++-v10/bits/invoke.h:73
    [...]
    (gdb) frame 1
    #1  0x00005555555ccbfd in cpr::HttpServer::initServer (this=0x5555556494c0, mgr=0x555555649568, event_handler=0x5555555ca35b <cpr::EventHandler(mg_connection*, int, void*)>)
        at /tmp/tmp.9PHOp0W6Cq/cpr/test/httpServer.cpp:19
    19          mg_set_protocol_http_websocket(c);
    (gdb) l
    14                                            MG_CB(mg_event_handler_t event_handler, void* user_data)) {
    15          // Based on: https://cesanta.com/docs/http/server-example.html
    16          mg_mgr_init(mgr, this);
    17          std::string port = std::to_string(GetPort());
    18          mg_connection* c = mg_bind(mgr, port.c_str(), event_handler);
    19          mg_set_protocol_http_websocket(c);
    20          return c;
    21      }
    22
    23      void HttpServer::OnRequestHello(mg_connection* conn, http_message* msg) {
    (gdb) p c
    $1 = (mg_connection *) 0x0
    (gdb) p mgr
    $2 = (mg_mgr *) 0x555555649568
    (gdb) p port
    $3 = "8080"
    (gdb) p event_handler
    $4 = (mg_event_handler_t) 0x5555555ca35b <cpr::EventHandler(mg_connection*, int, void*)>

**Comment by EmRowlands** [comment]
I think I have worked out the issue!

Port 8080 is already in use on my machine. Any time the test tries to bind to 8080, it fails. This could be solved by using a random port, or maybe a port in the ephemeral range?

**Comment by COM8** [comment]
Aha!
This makes sense. Yes, you are right, something like this should be done to prevent this from happening for other users.
Any chance you would like to create a PR for this?

**Comment by EmRowlands** [comment]
I created a small patch to change the ports used in the tests, but since the tests are run in parallel, this still causes problems when the tests are run in parallel (as is the case when tests are run with `-j2` etc.), so I have changed my configuration file to run tests linearly. It might be best to use CMakeLists to prevent running tests in parallel, but I don't know how to do this.

Unfortunately, I don't really know the best way of assigning random ports in this way, but I think the only thing that's really necessary is to check the return value of `mg_bind` and show an error if it fails. I'll submit a PR with my small patch today.
